src/url_search_params.rs

```diff
@@ -0,0 +1,483 @@
+use crate::{ffi, ParseUrlError};
+
+pub struct UrlSearchParams(*mut ffi::ada_url_search_params);
+
+impl Drop for UrlSearchParams {
+    fn drop(&mut self) {
+        unsafe { ffi::ada_free_search_params(self.0) }
+    }
+}
+
+impl UrlSearchParams {
+    /// Parses an return a UrlSearchParams struct.
+    ///
+    /// ```
+    /// use ada_url::UrlSearchParams;
+    /// let params = UrlSearchParams::parse("a=1&b=2")
+    ///     .expect("This is a valid UrlSearchParams. Should have parsed it.");
+    /// assert_eq!(params.get("a"), Some("1"));
+    /// assert_eq!(params.get("b"), Some("2"));
+    /// ```
+    pub fn parse<Input>(input: Input) -> Result<Self, ParseUrlError<Input>>
+    where
+        Input: AsRef<str>,
+    {
+        Ok(Self(unsafe {
+            ffi::ada_parse_search_params(input.as_ref().as_ptr().cast(), input.as_ref().len())
```

Comment:
This returns a `Result`, but can’t fail internally?

Does the underlying C function never return an error here? IMO its okay to just define returning an error for forwards compatibility so you can add it later, but I rather think there is something missing here?

Reply:
It doesn't fail yet, but I don't want to make a breaking change when the URLSearchParams c-api changes.